File: src/main/java/com/pixelreel/service/OpenXblImportService.java
package com.pixelreel.service;

import com.fasterxml.jackson.databind.JsonNode;
import com.pixelreel.config.OpenXblProperties;
import com.pixelreel.dto.imports.ImportSummary;
import com.pixelreel.entity.Game;
import com.pixelreel.enums.RecordStatus;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.function.Function;
import java.util.stream.Collectors;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.stereotype.Service;
import org.springframework.util.StringUtils;
import org.springframework.web.client.RestClient;

@Service
@RequiredArgsConstructor
public class OpenXblImportService {
  private static final String HEADER_API_KEY = "X-Authorization";

  private final GameService gameService;

  @Qualifier("openxblRestClient")
  private final RestClient openxblRestClient;

  private final OpenXblProperties openxblProperties;

  public ImportSummary importOwnedGames(String gamertag, RecordStatus status) {
    ImportSummary summary = new ImportSummary();
    summary.setErrors(new ArrayList<>());

    if (!openxblProperties.enabled()) {
      summary.getErrors().add("OpenXBL 未启用");
      return summary;
    }
    if (!StringUtils.hasText(openxblProperties.apiKey())) {
      summary.getErrors().add("缺少 OpenXBL API Key");
      return summary;
    }
    if (!StringUtils.hasText(gamertag)) {
      summary.getErrors().add("缺少 Xbox Gamertag");
      return summary;
    }

    JsonNode searchResponse = safeGet("/search/{gamertag}", gamertag.trim());
    String xuid = extractFirstText(searchResponse, "xuid");
    if (!StringUtils.hasText(xuid)) {
      summary.getErrors().add("无法从 OpenXBL 搜索结果解析 XUID");
      return summary;
    }

    JsonNode titleHistory = safeGet("/player/titleHistory/{xuid}", xuid);
    if (titleHistory == null || titleHistory.isMissingNode()) {
      summary.getErrors().add("获取 Xbox 游戏列表失败");
      return summary;
    }

    List<OpenXblTitle> titles = parseTitles(titleHistory);
    Map<String, AchievementStats> achievementStats = Collections.emptyMap();
    JsonNode achievementsResponse = safeGet("/achievements/player/{xuid}", xuid);
    if (achievementsResponse == null || achievementsResponse.isMissingNode()) {
      summary.getErrors().add("获取成就统计失败，已跳过成就数据");
    } else {
      achievementStats = parseAchievements(achievementsResponse);
    }

    Map<String, Game> existing = existingByXboxId(titles);
    List<Game> toSave = new ArrayList<>();
    RecordStatus effectiveStatus = status == null ? RecordStatus.UNSET : status;
    LocalDateTime now = LocalDateTime.now();

    summary.setTotal(titles.size());
    for (OpenXblTitle title : titles) {
      if (!StringUtils.hasText(title.titleId())) {
        summary.getErrors().add("缺少 Xbox titleId，已跳过: " + safeLabel(title.name()));
        summary.setSkipped(summary.getSkipped() + 1);
        continue;
      }
      if (!StringUtils.hasText(title.name())) {
        summary.getErrors().add("缺少游戏名称，已跳过: " + title.titleId());
        summary.setSkipped(summary.getSkipped() + 1);
        continue;
      }
      if (existing.containsKey(title.titleId())) {
        summary.setSkipped(summary.getSkipped() + 1);
        continue;
      }
      try {
        Game entity = new Game();
        entity.setXboxId(title.titleId());
        entity.setTitle(title.name());
        entity.setPosterUrl(title.posterUrl());
        entity.setPlatform("XBOX");
        entity.setPlaytimeMinutes(title.playtimeMinutes());
        entity.setStatus(effectiveStatus);
        entity.setRating(null);
        entity.setShortReview("");
        entity.setImportedAt(now);

        AchievementStats stats = achievementStats.get(title.titleId());
        if (stats != null) {
          entity.setAchievementTotal(stats.total());
          entity.setAchievementUnlocked(stats.unlocked());
        }

        toSave.add(entity);
      } catch (RuntimeException ex) {
        summary.getErrors().add("导入失败: " + safeLabel(title.name()) + "，原因: " + ex.getMessage());
        summary.setSkipped(summary.getSkipped() + 1);
      }
    }

    if (!toSave.isEmpty()) {
      gameService.saveBatch(toSave);
      summary.setImported(toSave.size());
    }
    return summary;
  }

  private JsonNode safeGet(String path, Object... uriVariables) {
    try {
      return openxblRestClient
          .get()
          .uri(path, uriVariables)
          .header(HEADER_API_KEY, openxblProperties.apiKey())
          .retrieve()
          .body(JsonNode.class);
    } catch (RuntimeException ex) {
      return null;
    }
  }

  private List<OpenXblTitle> parseTitles(JsonNode root) {
    if (root == null || root.isMissingNode()) {
      return Collections.emptyList();
    }

    JsonNode array = firstArray(root, List.of("titles", "titleHistory", "data", "items"));
    if (array == null || !array.isArray()) {
      array = findFirstArrayWithField(root, "titleId");
    }
    if (array == null || !array.isArray()) {
      return Collections.emptyList();
    }

    List<OpenXblTitle> titles = new ArrayList<>();
    for (JsonNode item : array) {
      String titleId = firstText(item, List.of("titleId", "titleID", "id", "title_id"));
      String name = firstText(item, List.of("name", "title", "displayName", "game"));
      String posterUrl = firstText(item, List.of("displayImage", "displayImageUrl", "image", "imageUrl", "boxArt", "boxArtUrl"));
      Integer playtimeMinutes = parsePlaytimeMinutes(item);
      if (StringUtils.hasText(titleId) || StringUtils.hasText(name)) {
        titles.add(new OpenXblTitle(titleId, name, posterUrl, playtimeMinutes));
      }
    }
    return titles;
  }

  private Map<String, AchievementStats> parseAchievements(JsonNode root) {
    if (root == null || root.isMissingNode()) {
      return Collections.emptyMap();
    }
    JsonNode array = firstArray(root, List.of("titles", "data", "items"));
    if (array == null || !array.isArray()) {
      array = findFirstArrayWithField(root, "titleId");
    }
    if (array == null || !array.isArray()) {
      return Collections.emptyMap();
    }

    Map<String, AchievementStats> map = new HashMap<>();
    for (JsonNode item : array) {
      String titleId = firstText(item, List.of("titleId", "titleID", "id", "title_id"));
      if (!StringUtils.hasText(titleId)) {
        continue;
      }

      Integer total = null;
      Integer unlocked = null;

      JsonNode statsNode = firstObject(item, List.of("achievement", "achievements", "stats", "progress"));
      if (statsNode != null && statsNode.isObject()) {
        total = firstInt(statsNode, List.of("totalAchievements", "total", "achievementTotal"));
        unlocked = firstInt(statsNode, List.of("currentAchievements", "unlockedAchievements", "achievementUnlocked", "earned"));
      }

      JsonNode achievementsArray = item.path("achievements");
      if ((total == null || unlocked == null) && achievementsArray.isArray()) {
        total = total == null ? achievementsArray.size() : total;
        if (unlocked == null) {
          int unlockedCount = 0;
          for (JsonNode achievement : achievementsArray) {
            if (isAchievementUnlocked(achievement)) {
              unlockedCount++;
            }
          }
          unlocked = unlockedCount;
        }
      }

      if (total != null || unlocked != null) {
        map.put(titleId, new AchievementStats(total, unlocked));
      }
    }
    return map;
  }

  private boolean isAchievementUnlocked(JsonNode achievement) {
    if (achievement == null || achievement.isMissingNode()) {
      return false;
    }
    JsonNode progressState = achievement.path("progressState");
    if (progressState.isTextual()) {
      String normalized = progressState.asText("").trim().toLowerCase(Locale.ROOT);
      if (normalized.contains("achieved") || normalized.contains("unlocked") || normalized.contains("completed")) {
        return true;
      }
    }
    JsonNode unlocked = achievement.path("unlocked");
    if (unlocked.isBoolean()) {
      return unlocked.asBoolean();
    }
    JsonNode progress = achievement.path("progress");
    if (progress.isNumber()) {
      return progress.asInt() >= 100;
    }
    return false;
  }

  private Integer parsePlaytimeMinutes(JsonNode item) {
    Integer minutes = firstInt(item, List.of("playtimeMinutes", "minutesPlayed", "timePlayedMinutes", "minutes"));
    if (minutes != null) {
      return minutes;
    }
    Integer hours = firstInt(item, List.of("playtimeHours", "hoursPlayed", "timePlayedHours", "hours"));
    if (hours != null) {
      return hours * 60;
    }

    String timeText = firstText(item, List.of("timePlayed", "playtime"));
    if (!StringUtils.hasText(timeText)) {
      return null;
    }
    String normalized = timeText.trim().toLowerCase(Locale.ROOT);
    Double numeric = parseLeadingNumber(normalized);
    if (numeric == null) {
      return null;
    }
    if (normalized.contains("hour") || normalized.endsWith("h")) {
      return (int) Math.round(numeric * 60);
    }
    if (normalized.contains("min") || normalized.endsWith("m")) {
      return (int) Math.round(numeric);
    }
    return (int) Math.round(numeric);
  }

  private Double parseLeadingNumber(String value) {
    StringBuilder builder = new StringBuilder();
    for (int i = 0; i < value.length(); i++) {
      char ch = value.charAt(i);
      if (Character.isDigit(ch) || ch == '.') {
        builder.append(ch);
      } else if (builder.length() > 0) {
        break;
      }
    }
    if (builder.length() == 0) {
      return null;
    }
    try {
      return Double.parseDouble(builder.toString());
    } catch (NumberFormatException ex) {
      return null;
    }
  }

  private Map<String, Game> existingByXboxId(List<OpenXblTitle> titles) {
    List<String> ids = titles.stream()
        .map(OpenXblTitle::titleId)
        .filter(StringUtils::hasText)
        .distinct()
        .toList();
    if (ids.isEmpty()) {
      return Collections.emptyMap();
    }
    return gameService.lambdaQuery()
        .in(Game::getXboxId, ids)
        .list()
        .stream()
        .collect(Collectors.toMap(Game::getXboxId, Function.identity(), (a, b) -> a));
  }

  private String extractFirstText(JsonNode root, String fieldName) {
    if (root == null || root.isMissingNode()) {
      return null;
    }
    if (root.has(fieldName)) {
      JsonNode value = root.get(fieldName);
      if (value.isTextual() || value.isNumber()) {
        return value.asText();
      }
    }
    if (root.isArray()) {
      for (JsonNode child : root) {
        String found = extractFirstText(child, fieldName);
        if (StringUtils.hasText(found)) {
          return found;
        }
      }
      return null;
    }
    if (root.isObject()) {
      for (JsonNode child : root) {
        String found = extractFirstText(child, fieldName);
        if (StringUtils.hasText(found)) {
          return found;
        }
      }
    }
    return null;
  }

  private JsonNode firstArray(JsonNode root, List<String> candidates) {
    for (String candidate : candidates) {
      JsonNode node = root.path(candidate);
      if (node.isArray()) {
        return node;
      }
    }
    return null;
  }

  private JsonNode findFirstArrayWithField(JsonNode root, String fieldName) {
    if (root == null || root.isMissingNode()) {
      return null;
    }
    if (root.isArray()) {
      for (JsonNode child : root) {
        if (child.isObject() && child.has(fieldName)) {
          return root;
        }
      }
      for (JsonNode child : root) {
        JsonNode found = findFirstArrayWithField(child, fieldName);
        if (found != null) {
          return found;
        }
      }
      return null;
    }
    if (root.isObject()) {
      for (JsonNode child : root) {
        JsonNode found = findFirstArrayWithField(child, fieldName);
        if (found != null) {
          return found;
        }
      }
    }
    return null;
  }

  private JsonNode firstObject(JsonNode root, List<String> candidates) {
    for (String candidate : candidates) {
      JsonNode node = root.path(candidate);
      if (node.isObject()) {
        return node;
      }
    }
    return null;
  }

  private String firstText(JsonNode root, List<String> candidates) {
    for (String candidate : candidates) {
      JsonNode node = root.path(candidate);
      if (node.isTextual()) {
        String value = node.asText();
        if (StringUtils.hasText(value)) {
          return value.trim();
        }
      }
      if (node.isNumber()) {
        return node.asText();
      }
    }
    return null;
  }

  private Integer firstInt(JsonNode root, List<String> candidates) {
    for (String candidate : candidates) {
      JsonNode node = root.path(candidate);
      if (node.isNumber()) {
        return node.intValue();
      }
      if (node.isTextual()) {
        String text = node.asText();
        if (StringUtils.hasText(text)) {
          try {
            return (int) Math.round(Double.parseDouble(text.trim()));
          } catch (NumberFormatException ignored) {
            // ignore
          }
        }
      }
    }
    return null;
  }

  private String safeLabel(String value) {
    return StringUtils.hasText(value) ? value : "未知游戏";
  }

  private record OpenXblTitle(String titleId, String name, String posterUrl, Integer playtimeMinutes) {}

  private record AchievementStats(Integer total, Integer unlocked) {}
}

File: express-backend/src/services/import/OpenXblImportService.ts
import axios from 'axios';
import { config } from '../../config';
import { getDb } from '../../config/db';
import { ImportSummary } from '../../dto/import-summary';
import { RecordStatus } from '../../enums/RecordStatus';
import { lookupRawgPosterUrl } from './RawgPosterLookupService';
import {
  buildPlatformGameRequestOptions,
  buildPlatformGameMetricUpdate,
  hasPlatformGameMetricUpdate,
  isPlatformGameExternalIdValid,
  isPlatformGameTitleValid,
  normalizePlatformGamePosterUrl,
  parsePlatformGameMetric,
  PLATFORM_GAME_EXTERNAL_ID_MAX_LENGTH,
  PLATFORM_GAME_TITLE_MAX_LENGTH,
} from './PlatformGameSyncService';

export interface XboxImportedTitle {
  titleId: string;
  name: string | null;
  posterUrl: string | null;
  playtimeMinutes: number | null;
  achievementTotal: number | null;
  achievementUnlocked: number | null;
}

type ImportProgress = (processed: number, total: number, currentTitle: string) => void;

// Xbox 已玩游戏导入服务，与 Java 端 OpenXblImportService 完全对齐
export async function importXboxOwnedGames(
  gamertag: string,
  status?: string | null,
  onProgress?: ImportProgress,
  signal?: AbortSignal,
): Promise<ImportSummary> {
  const summary: ImportSummary = { total: 0, imported: 0, updated: 0, skipped: 0, errors: [] };

  if (!config.openxbl.enabled) {
    summary.errors.push('OpenXBL 未启用');
    return summary;
  }
  if (!config.openxbl.apiKey) {
    summary.errors.push('缺少 OpenXBL API Key');
    return summary;
  }
  if (!gamertag) {
    summary.errors.push('缺少 Xbox Gamertag');
    return summary;
  }

  // 1. 通过 gamertag 查 XUID
  let xuid: string | null = null;
  onProgress?.(0, 0, '解析 Xbox 账号');
  try {
    const searchRes = await axios.get(buildOpenXblUrl(
      config.openxbl.baseUrl,
      `search/${encodeURIComponent(gamertag.trim())}`,
    ), {
      headers: { 'X-Authorization': config.openxbl.apiKey },
      ...buildPlatformGameRequestOptions(signal),
    });
    xuid = extractXuid(searchRes.data, gamertag.trim());
  } catch (ex: any) {
    if (signal?.aborted) return summary;
    summary.errors.push(`Xbox 用户搜索失败: ${ex.message}`);
    return summary;
  }

  if (!xuid) {
    summary.errors.push('OpenXBL 搜索结果中没有完全匹配的 Gamertag');
    return summary;
  }

  // 2. 获取游戏列表
  let titleHistory: XboxImportedTitle[] = [];
  try {
    onProgress?.(0, 0, '读取 Xbox 游戏库');
    const titleRes = await axios.get(buildOpenXblUrl(
      config.openxbl.baseUrl,
      `player/titleHistory/${encodeURIComponent(xuid)}`,
    ), {
      headers: { 'X-Authorization': config.openxbl.apiKey },
      ...buildPlatformGameRequestOptions(signal),
    });
    titleHistory = parseXboxTitles(titleRes.data);
  } catch (ex: any) {
    if (signal?.aborted) return summary;
    summary.errors.push(`获取 Xbox 游戏列表失败: ${ex.message}`);
    return summary;
  }

  summary.total = titleHistory.length;
  const effectiveStatus = status || RecordStatus.UNSET;
  const now = new Date();

  // 批量查已有记录
  const xboxIds = titleHistory.map((t) => t.titleId).filter(Boolean);
  const existingMap = xboxIds.length > 0
    ? new Map((await getDb().game.findMany({ where: { xboxId: { in: xboxIds } } })).map((g) => [g.xboxId!, g]))
    : new Map<any, any>();

  const toSave: any[] = [];
  for (const [index, title] of titleHistory.entries()) {
    if (signal?.aborted) return summary;
    onProgress?.(index + 1, summary.total, title.name || title.titleId);
    if (!title.titleId) {
      summary.errors.push(`缺少 Xbox titleId，已跳过: ${title.name || '未知游戏'}`);
      summary.skipped++;
      continue;
    }
    if (!isPlatformGameExternalIdValid(title.titleId)) {
      summary.errors.push(`Xbox titleId 超过 ${PLATFORM_GAME_EXTERNAL_ID_MAX_LENGTH} 个字符，已跳过: ${title.name || '未知游戏'}`);
      summary.skipped++;
      continue;
    }
    if (!title.name) {
      summary.errors.push(`缺少游戏名称，已跳过: ${title.titleId}`);
      summary.skipped++;
      continue;
    }
    if (!isPlatformGameTitleValid(title.name)) {
      summary.errors.push(`Xbox 游戏标题超过 ${PLATFORM_GAME_TITLE_MAX_LENGTH} 个字符，已跳过: ${title.titleId}`);
      summary.skipped++;
      continue;
    }
    const sourcePosterUrl = normalizePlatformGamePosterUrl(title.posterUrl);
    if (title.posterUrl && !sourcePosterUrl) {
      summary.errors.push(`Xbox 封面 URL 超过字段长度，已改用回退封面: ${title.name}`);
    }
    const existing = existingMap.get(title.titleId);
    if (existing) {
      const posterUrl = existing.posterUrl
        ? null
        : sourcePosterUrl ?? await lookupRawgPosterUrl(title.name, signal);
      if (signal?.aborted) return summary;
      const update = buildPlatformGameMetricUpdate(existing, {
        platform: 'XBOX',
        posterUrl,
        playtimeMinutes: title.playtimeMinutes,
        achievementTotal: title.achievementTotal,
        achievementUnlocked: title.achievementUnlocked,
      });
      if (!hasPlatformGameMetricUpdate(update)) {
        summary.skipped++;
        continue;
      }
      await getDb().game.update({ where: { id: existing.id }, data: update });
      summary.updated = (summary.updated ?? 0) + 1;
      continue;
    }

    const posterUrl = sourcePosterUrl
      ?? await lookupRawgPosterUrl(title.name, signal);
    if (signal?.aborted) return summary;

    toSave.push({
      xboxId: title.titleId,
      title: title.name,
      posterUrl,
      platform: 'XBOX',
      playtimeMinutes: title.playtimeMinutes,
      achievementTotal: title.achievementTotal,
      achievementUnlocked: title.achievementUnlocked,
      importedAt: now,
      importReviewState: 'PENDING',
      status: effectiveStatus,
      rating: null,
      shortReview: '',
    });
  }

  if (signal?.aborted) return summary;
  if (toSave.length > 0) {
    await getDb().game.createMany({ data: toSave });
    summary.imported = toSave.length;
  }
  onProgress?.(summary.total, summary.total, '');

  return summary;
}

export function extractXuid(data: unknown, expectedGamertag?: string): string | null {
  const normalizedExpected = expectedGamertag
    ? normalizeXboxGamertag(expectedGamertag)
    : null;
  return findXboxXuid(data, normalizedExpected);
}

function findXboxXuid(data: unknown, normalizedExpected: string | null): string | null {
  if (!data || typeof data !== 'object') return null;
  if (Array.isArray(data)) {
    for (const item of data) {
      const found = findXboxXuid(item, normalizedExpected);
      if (found) return found;
    }
    return null;
  }

  const record = data as Record<string, unknown>;
  const modernGamertag = readString(record.modernGamertag ?? record.ModernGamertag);
  const modernGamertagSuffix = readString(
    record.modernGamertagSuffix ?? record.ModernGamertagSuffix,
  );
  const returnedGamertags = [
    readString(record.gamertag ?? record.Gamertag),
    readString(record.uniqueModernGamertag ?? record.UniqueModernGamertag),
    modernGamertag,
    modernGamertag && modernGamertagSuffix
      ? `${modernGamertag}#${modernGamertagSuffix}`
      : null,
  ].filter((value): value is string => value != null);
  if (normalizedExpected == null
    || returnedGamertags.some(gamertag => normalizeXboxGamertag(gamertag) === normalizedExpected)) {
    const xuid = parseXboxXuid(record.xuid) ?? parseXboxXuid(record.Xuid);
    if (xuid) return xuid;
  }
  for (const value of Object.values(record)) {
    const found = findXboxXuid(value, normalizedExpected);
    if (found) return found;
  }
  return null;
}

function normalizeXboxGamertag(value: string): string {
  return value.trim().normalize('NFKC').toLocaleLowerCase('en-US');
}

export function buildOpenXblUrl(baseUrl: string, resourcePath: string): string {
  return `${baseUrl.replace(/\/+$/, '')}/${resourcePath.replace(/^\/+/, '')}`;
}

function parseXboxXuid(value: unknown): string | null {
  if (typeof value === 'number' && (!Number.isSafeInteger(value) || value <= 0)) return null;
  if (typeof value !== 'string' && typeof value !== 'number') return null;
  const normalized = String(value).trim();
  return /^[1-9]\d{0,19}$/.test(normalized) ? normalized : null;
}

export function parseXboxTitles(data: unknown): XboxImportedTitle[] {
  const array = findXboxTitleArray(data);
  if (!array) throw new Error('OpenXBL 返回的游戏列表格式无效');
  const seenTitleIds = new Set<string>();
  return array.flatMap((value) => {
    if (!value || typeof value !== 'object') return [];
    const item = value as Record<string, unknown>;
    const type = readString(item.type);
    if (type && type.toLowerCase() !== 'game') return [];

    const titleId = readString(item.titleId ?? item.titleID ?? item.id ?? item.title_id);
    if (!titleId || seenTitleIds.has(titleId)) return [];
    seenTitleIds.add(titleId);
    const statsValue = item.achievement ?? item.achievements ?? item.stats ?? item.progress;
    const stats = statsValue && typeof statsValue === 'object'
      ? statsValue as Record<string, unknown>
      : {};
    return [{
      titleId,
      name: readString(item.name ?? item.title ?? item.displayName ?? item.game),
      posterUrl: readString(
        item.displayImage ?? item.displayImageUrl ?? item.image
        ?? item.imageUrl ?? item.boxArt ?? item.boxArtUrl,
      ),
      playtimeMinutes: parsePlatformGameMetric(
        item.playtimeMinutes ?? item.minutesPlayed ?? item.timePlayedMinutes ?? item.minutes,
      ),
      achievementTotal: parsePlatformGameMetric(
        stats.totalAchievements ?? stats.total ?? stats.achievementTotal,
      ),
      achievementUnlocked: parsePlatformGameMetric(
        stats.currentAchievements ?? stats.unlockedAchievements
        ?? stats.achievementUnlocked ?? stats.earned,
      ),
    }];
  });
}

function findXboxTitleArray(data: unknown): unknown[] | null {
  if (!data || typeof data !== 'object') return null;
  const record = data as Record<string, unknown>;
  const content = record.content && typeof record.content === 'object'
    ? record.content as Record<string, unknown>
    : record;
  const direct = content.titles ?? content.titleHistory ?? content.data ?? content.items;
  if (Array.isArray(direct)) return direct;
  for (const value of Object.values(content)) {
    if (Array.isArray(value) && value.some(item => (
      item && typeof item === 'object'
      && ('titleId' in item || 'titleID' in item)
    ))) return value;
  }
  return null;
}

function readString(value: unknown): string | null {
  if (typeof value !== 'string' && typeof value !== 'number') return null;
  const normalized = String(value).trim();
  return normalized || null;
}
